From 9ac92eb18292b5005e43f1bd60d7e4a160ddb997 Mon Sep 17 00:00:00 2001
From: feeRnt <81442162+feeRnt@users.noreply.github.com>
Date: Thu, 21 Aug 2025 18:58:02 +0000
Subject: [PATCH] drm: amdgpu - Fix blackscreen at Display Manager

Should fix the blackscreen that occurs when going from kernel
and initramfs to a distro's graphical display manager.
Makes the switch / refresh to a new mode not cause the blackscreen.

Files changed:

- drivers/gpu/drm/amd/amdgpu/ps4_bridge.c
Removed cq_delay(s) in ps4_bridge.c.
PS4-Linux patches from github.com/rancido; part 1.

drm_bridge_attach without DRM_BRIDGE_ATTACH_NO_CONNECTOR (with 0 now).
PS4-Linux patches from github.com/rancido; part 5.

Signed-off-by: feeRnt <81442162+feeRnt@users.noreply.github.com>
---
 drivers/gpu/drm/amd/amdgpu/ps4_bridge.c | 33 ++++++++++++++++++++-----
 1 file changed, 27 insertions(+), 6 deletions(-)

diff --git a/drivers/gpu/drm/amd/amdgpu/ps4_bridge.c b/drivers/gpu/drm/amd/amdgpu/ps4_bridge.c
index ca35b66efc19b80e1a1b29b88725840f5ea75b09..bd50c960280f5cf0c5033a50477edb10ebf6598a 100644
--- a/drivers/gpu/drm/amd/amdgpu/ps4_bridge.c
+++ b/drivers/gpu/drm/amd/amdgpu/ps4_bridge.c
@@ -1,5 +1,5 @@
 /*
- * Panasonic MN86471A DP->HDMI bridge driver (via PS4 Aeolia ICC interface)
+ * Panasonic MN86471A / MN864729 DP->HDMI bridge driver (via PS4 Aeolia ICC interface)
  *
  * This software is licensed under the terms of the GNU General Public
  * License version 2, as published by the Free Software Foundation, and
@@ -15,6 +15,20 @@
 // TODO (ps4patches): Make functions atomic,
 //  https://lore.kernel.org/linux-arm-kernel/20211020181901.2114645-5-sam@ravnborg.org/
 
+/*
+ * Forward-ported patches from https://github.com/rancido,
+ * 		            in https://github.com/Ps3itaTeam/ps4-linux/
+ * Maybe the problem is no one is putting a "Copyright"
+ * in their files ... We don't know who introduced what change or where,
+ * since there is often no centralized repository.. Patches thrown around
+ * everywhere.
+ * But that goes against the whole nature of tinkery/RE work..?
+ * But does it really? No.
+ *
+ * 
+ * Copyright: .....
+ *
+*/
 #include <asm/ps4.h>
 
 #include <drm/drm_crtc.h>
@@ -247,7 +261,7 @@ static void cq_mask(struct i2c_cmdqueue *q, u16 addr, u8 value, u8 mask)
 	*q->p++ = mask;
 }
 
-#if 1
+#if 0
 static void cq_delay(struct i2c_cmdqueue *q, u16 time)
 {
 	cq_cmd(q, CMD_DELAY);
@@ -524,9 +538,12 @@ static void ps4_bridge_enable(struct drm_bridge *bridge)
 		cq_mask(&mn_bridge->cq, 0x1e00, 0x00, 0x21);
 		cq_mask(&mn_bridge->cq, 0x1e02, 0x00, 0x70);
 		// 03 08 01 01 00  2c 01 00
-		cq_delay(&mn_bridge->cq, 0x012c);
+		//rancido has no delay here vvv
+		//cq_delay(&mn_bridge->cq, 0x012c);
 		cq_writereg(&mn_bridge->cq, 0x6020, 0x00);
-		cq_delay(&mn_bridge->cq, 0x0032);
+
+		//rancido has no delay here vvv
+		//cq_delay(&mn_bridge->cq, 0x0032);
 		cq_writereg(&mn_bridge->cq, 0x7402, 0x1c);
 		cq_writereg(&mn_bridge->cq, 0x6020, 0x04);
 		cq_writereg(&mn_bridge->cq, TSYSCTRL, TSYSCTRL_HDMI);
@@ -569,7 +586,8 @@ static void ps4_bridge_enable(struct drm_bridge *bridge)
 		cq_wait_set(&mn_bridge->cq, 0x10f6, 0x80);
 		cq_mask(&mn_bridge->cq, 0x7226, 0x00, 0x80);
 		cq_mask(&mn_bridge->cq, 0x7228, 0x00, 0xFF);
-		cq_delay(&mn_bridge->cq, 0x012c);
+		// rancido has no delay here vvv
+		//cq_delay(&mn_bridge->cq, 0x012c);
 		cq_writereg(&mn_bridge->cq, 0x7204, 0x40);
 		cq_wait_clear(&mn_bridge->cq, 0x7204, 0x40);
 		cq_writereg(&mn_bridge->cq, 0x7a8b, 0x05);
@@ -780,7 +798,10 @@ int ps4_bridge_register(struct drm_connector *connector,
 	// TODO (ps4patches): This seems to be the new way of adding bridges
 	drm_bridge_add(&mn_bridge->bridge);
 
-	ret = drm_bridge_attach(mn_bridge->encoder, &mn_bridge->bridge, NULL, DRM_BRIDGE_ATTACH_NO_CONNECTOR);
+	//ret = drm_bridge_attach(mn_bridge->encoder, &mn_bridge->bridge, NULL, DRM_BRIDGE_ATTACH_NO_CONNECTOR);
+	
+	// Allow the bridge to create its own connectors with last parameter = 0.
+	ret = drm_bridge_attach(mn_bridge->encoder, &mn_bridge->bridge, NULL, 0);
 	if (ret) {
 		DRM_ERROR("Failed to initialize bridge with drm\n");
 		return -EINVAL;
